### PR TITLE
Added ability to use YAML instead

### PR DESCRIPTION
**Pull request by ggstrader**
I've updated the dependencies to get rid of vulnerabilities and use the updated obsidian API

I've also added the option to insert the classes in the YAML frontmatter rather than just modifying the elements. This allows for better future-proofing and interoperability with other plugins.

### CODE REVIEW

File: src/settings/settings.ts [hunk 1]
```diff
@@ -147,6 +151,59 @@ export class AutoClassPluginSettingsTab extends PluginSettingTab {
...
+        text.setPlaceholder(DEFAULT_SETTINGS.yamlAttribute).onChange(async (value) => {
+          // only allow letters, numbers, dashes, and underscores
+          this.plugin.settings.yamlAttribute = value.replace(/[^a-zA-Z0-9-_]/g, '') || null;
+          // if we changed the attribute name, we need to update settings view
+          if (value && this.plugin.settings.yamlAttribute !== value) {
+            text.setValue(this.plugin.settings.yamlAttribute);
+            return;
+          }
+          // otherwise use the default if the user cleared the field
+          this.plugin.settings.yamlAttribute ??= DEFAULT_SETTINGS.yamlAttribute;
```
Comment (OfficerHalf) [inline src/settings/settings.ts:202]: That's some fancy syntax.

**Review by nathonius — COMMENTED** [review]
@ggstrader hey, thanks for the contribution! The code itself looks good to me, however this does conflict a little with some plans I had to overhaul this plugin a bit. I definitely want to keep the new option to add classes to YAML rather than on the container, but I want to do that at a rule level rather than globally; the original intent of this plugin was to allow using different vault themes depending on the file/folder you were viewing. Special file formatting instead is good, but I want it to be per-rule.

All that to say - I'm not going to merge this quite yet, but I'll likely use this as the starting point for adding this feature soon as part of the aforementioned overhaul.

**Comment by ggstrader** [comment]
> @ggstrader hey, thanks for the contribution! The code itself looks good to me, however this does conflict a little with some plans I had to overhaul this plugin a bit. I definitely want to keep the new option to add classes to YAML rather than on the container, but I want to do that at a rule level rather than globally; the original intent of this plugin was to allow using different vault themes depending on the file/folder you were viewing. Special file formatting instead is good, but I want it to be per-rule.
> 
> All that to say - I'm not going to merge this quite yet, but I'll likely use this as the starting point for adding this feature soon as part of the aforementioned overhaul.

Maybe I'm misunderstanding, but I believe it does what you're saying? It assigns classes based on the rules, not globally. It's a one to one with the old feature set, it just adds and removes to YAML based on the rules, which allows multiple classes per file as before.

File: src/plugin.ts [hunk 2]
```diff
@@ -40,17 +41,13 @@ export class AutoClassPlugin extends Plugin {
...
-      let matches: Array<ClassPath | ClassTag> = [];
-      let container: Element;
-      if (this.isReadMode(view)) {
-        matches = this.getMatches(view, allClasses, ClassMatchScope.Read);
-        container = this.getPreviewContainer(view);
-      } else if (this.isEditMode(view)) {
-        matches = this.getMatches(view, allClasses, ClassMatchScope.Edit);
-        container = this.getEditContainer(view);
-      }
+      const [getContainer, scope] = this.isReadMode(view)
```
[the remark this answers is not in the document]
Comment (ggstrader) [reply src/plugin.ts:44]: I typically would as well, I just tend to refactor things I have to double-take on. At first I wasn't sure what the difference was between the conditions, I thought this way made it more obvious and maintainable, but that might just be me.

File: src/plugin.ts [hunk 3]
```diff
@@ -127,8 +124,14 @@ export class AutoClassPlugin extends Plugin {
    * they were added to for removal later.
    */
   private applyClasses(classes: string[], view: MarkdownView, container: Element): void {
-    container.addClasses(classes);
     this.appliedClasses.set(view, classes);
+    if (!classes?.length) return;
+    if (this.settings.writeToYAML)
+      this.app.fileManager.processFrontMatter(view.file, (fm) => {
+        fm[this.settings.yamlAttribute] = [...new Set([...classes, ...(fm[this.settings.yamlAttribute] || [])])];
+        return;
```
[the remark this answers is not in the document]
Comment (ggstrader) [reply src/plugin.ts:132]: Oh, I see now what you meant in the other comment. That could be useful for people, I agree.

**Comment by ggstrader** [comment]
Nevermind, I read the other comment and I see what you mean.

**Pull request closed without merging**